File: scripts/conflict_resolution.py
#!/usr/bin/env python3
"""
conflict_resolution.py

A foundational skeleton for automated Git conflict resolution handlers.
As per Omni-Workspace directives, this script must prioritize feature preservation.
It currently analyzes conflict markers but delegates actual merging to LLM agents
until full programmatic self-healing is implemented.
"""

import os
import subprocess
import sys

def check_for_conflicts(repo_path):
    """
    Returns a list of files currently in a conflicted state in the given repository.
    """
    try:
        # --name-only and --diff-filter=U gets unmerged (conflicted) files
        result = subprocess.run(
            ["git", "diff", "--name-only", "--diff-filter=U"],
            cwd=repo_path, check=True, text=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
        )
        files = result.stdout.strip().splitlines()
        return [f for f in files if f]
    except subprocess.CalledProcessError:
        return []

def extract_conflict_blocks(file_path):
    """
    Parses a file and extracts standard Git conflict blocks.
    Returns a list of dictionaries containing the 'local' and 'remote' changes.
    """
    if not os.path.exists(file_path):
        return []

    conflicts = []
    current_conflict = None
    state = "normal" # States: normal, local, remote

    with open(file_path, "r", encoding="utf-8", errors="ignore") as f:
        for line in f:
            if line.startswith("<<<<<<< "):
                current_conflict = {"local": [], "remote": []}
                state = "local"
            elif line.startswith("======="):
                if state == "local":
                    state = "remote"
            elif line.startswith(">>>>>>> "):
                if current_conflict and state == "remote":
                    conflicts.append(current_conflict)
                    current_conflict = None
                    state = "normal"
            else:
                if state == "local" and current_conflict:
                    current_conflict["local"].append(line)
                elif state == "remote" and current_conflict:
                    current_conflict["remote"].append(line)

    return conflicts

def analyze_conflict(conflict_block):
    """
    Analyzes a single conflict block and suggests a resolution strategy.
    Rule: Never Lose Features.
    """
    local_content = "".join(conflict_block["local"])
    remote_content = "".join(conflict_block["remote"])

    if not local_content.strip() and remote_content.strip():
         return "Accept Remote (Local is empty)"
    if local_content.strip() and not remote_content.strip():
         return "Accept Local (Remote is empty)"

    # If both have content, we defer to intelligent merging (LLM/Manual)
    return "Requires Intelligent Merge (Both contain features)"

def main():
    repo_path = os.getcwd()
    if not os.path.exists(os.path.join(repo_path, ".git")):
        print("Error: Must be run from the root of a git repository.")
        sys.exit(1)

    print(f"Scanning for conflicts in {repo_path}...")
    conflicted_files = check_for_conflicts(repo_path)

    if not conflicted_files:
        print("No conflicts detected.")
        return

    for file in conflicted_files:
        print(f"\n--- Conflicted File: {file} ---")
        full_path = os.path.join(repo_path, file)
        blocks = extract_conflict_blocks(full_path)

        for i, block in enumerate(blocks):
            strategy = analyze_conflict(block)
            print(f"  Conflict Block #{i+1}: {strategy}")

    print("\nWARNING: Automated resolution is in Skeleton phase. Manual LLM intervention required for Intelligent Merges.")

if __name__ == "__main__":
    main()

File: backend-go/conflict_resolution.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"strings"
)

type ConflictBlock struct {
	Local  []string `json:"local"`
	Remote []string `json:"remote"`
}

type ConflictReport struct {
	File     string          `json:"file"`
	Blocks   []ConflictBlock `json:"blocks"`
	Strategy []string        `json:"strategy"`
}

func checkForConflicts(repoPath string) ([]string, error) {
	cmd := exec.Command("git", "diff", "--name-only", "--diff-filter=U")
	cmd.Dir = repoPath
	out, err := cmd.Output()
	if err != nil {
		return nil, err
	}

	lines := strings.Split(strings.TrimSpace(string(out)), "\n")
	var files []string
	for _, f := range lines {
		if f != "" {
			files = append(files, f)
		}
	}
	return files, nil
}

func extractConflictBlocks(filePath string) ([]ConflictBlock, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var conflicts []ConflictBlock
	var currentConflict *ConflictBlock
	state := "normal" // normal, local, remote

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()

		if strings.HasPrefix(line, "<<<<<<< ") {
			currentConflict = &ConflictBlock{}
			state = "local"
		} else if strings.HasPrefix(line, "=======") {
			if state == "local" {
				state = "remote"
			}
		} else if strings.HasPrefix(line, ">>>>>>> ") {
			if currentConflict != nil && state == "remote" {
				conflicts = append(conflicts, *currentConflict)
				currentConflict = nil
				state = "normal"
			}
		} else {
			if state == "local" && currentConflict != nil {
				currentConflict.Local = append(currentConflict.Local, line)
			} else if state == "remote" && currentConflict != nil {
				currentConflict.Remote = append(currentConflict.Remote, line)
			}
		}
	}

	return conflicts, scanner.Err()
}

func analyzeConflict(block ConflictBlock) string {
	localContent := strings.TrimSpace(strings.Join(block.Local, ""))
	remoteContent := strings.TrimSpace(strings.Join(block.Remote, ""))

	if localContent == "" && remoteContent != "" {
		return "Accept Remote (Local is empty)"
	}
	if localContent != "" && remoteContent == "" {
		return "Accept Local (Remote is empty)"
	}

	return "Requires Intelligent Merge (Both contain features)"
}

func conflictResolutionHandler(w http.ResponseWriter, r *http.Request) {
	repoPath := ".." // Root of the repository from backend-go

	files, err := checkForConflicts(repoPath)
	if err != nil {
		http.Error(w, fmt.Sprintf("Failed to check for conflicts: %v", err), http.StatusInternalServerError)
		return
	}

	var reports []ConflictReport
	for _, file := range files {
		fullPath := repoPath + "/" + file
		blocks, err := extractConflictBlocks(fullPath)
		if err != nil {
			continue // Skip files that can't be read
		}

		var strategies []string
		for _, block := range blocks {
			strategies = append(strategies, analyzeConflict(block))
		}

		reports = append(reports, ConflictReport{
			File:     file,
			Blocks:   blocks,
			Strategy: strategies,
		})
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(reports)
}
